File: backend/src/services/userService.ts
import { query } from '../config/database';
import { User } from '../types';
import { AppError } from '../middleware/errorHandler';
import { logger } from '../utils/logger';

export interface UpdateProfileData {
  displayName?: string;
  bio?: string;
  avatarUrl?: string;
  bannerUrl?: string;
  sportCategories?: string[];
}

export class UserService {
  async getUserById(userId: string): Promise<any> {
    const result = await query(
      `SELECT u.id, u.username, u.role, u.display_name, u.bio, u.avatar_url, u.banner_url,
              u.created_at, u.is_verified,
              ap.sport_categories, ap.follower_count, ap.total_view_count, ap.is_live
       FROM users u
       LEFT JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE u.id = $1 AND u.is_banned = FALSE`,
      [userId]
    );

    if (result.rows.length === 0) {
      throw new AppError('User not found', 404);
    }

    return result.rows[0];
  }

  async getUserByUsername(username: string): Promise<any> {
    const result = await query(
      `SELECT u.id, u.username, u.role, u.display_name, u.bio, u.avatar_url, u.banner_url,
              u.created_at, u.is_verified,
              ap.sport_categories, ap.follower_count, ap.total_view_count, ap.is_live
       FROM users u
       LEFT JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE u.username = $1 AND u.is_banned = FALSE`,
      [username]
    );

    if (result.rows.length === 0) {
      throw new AppError('User not found', 404);
    }

    return result.rows[0];
  }

  async updateProfile(userId: string, data: UpdateProfileData): Promise<any> {
    const updates: string[] = [];
    const values: any[] = [];
    let paramCounter = 1;

    if (data.displayName !== undefined) {
      updates.push(`display_name = $${paramCounter++}`);
      values.push(data.displayName);
    }

    if (data.bio !== undefined) {
      updates.push(`bio = $${paramCounter++}`);
      values.push(data.bio);
    }

    if (data.avatarUrl !== undefined) {
      updates.push(`avatar_url = $${paramCounter++}`);
      values.push(data.avatarUrl);
    }

    if (data.bannerUrl !== undefined) {
      updates.push(`banner_url = $${paramCounter++}`);
      values.push(data.bannerUrl);
    }

    if (updates.length > 0) {
      values.push(userId);
      await query(
        `UPDATE users SET ${updates.join(', ')}, updated_at = CURRENT_TIMESTAMP
         WHERE id = $${paramCounter}`,
        values
      );
    }

    // Update athlete profile if sport categories provided
    if (data.sportCategories !== undefined) {
      await query(
        'UPDATE athlete_profiles SET sport_categories = $1 WHERE user_id = $2',
        [data.sportCategories, userId]
      );
    }

    logger.info(`Profile updated for user: ${userId}`);

    return this.getUserById(userId);
  }

  async followUser(followerId: string, athleteId: string): Promise<void> {
    // Check if athlete exists and is an athlete
    const athleteResult = await query(
      'SELECT role FROM users WHERE id = $1',
      [athleteId]
    );

    if (athleteResult.rows.length === 0) {
      throw new AppError('Athlete not found', 404);
    }

    if (athleteResult.rows[0].role !== 'athlete') {
      throw new AppError('Can only follow athletes', 400);
    }

    if (followerId === athleteId) {
      throw new AppError('Cannot follow yourself', 400);
    }

    // Check if already following
    const existingFollow = await query(
      'SELECT 1 FROM follows WHERE follower_id = $1 AND athlete_id = $2',
      [followerId, athleteId]
    );

    if (existingFollow.rows.length > 0) {
      throw new AppError('Already following this athlete', 400);
    }

    // Create follow relationship
    await query(
      'INSERT INTO follows (follower_id, athlete_id) VALUES ($1, $2)',
      [followerId, athleteId]
    );

    logger.info(`User ${followerId} followed athlete ${athleteId}`);
  }

  async unfollowUser(followerId: string, athleteId: string): Promise<void> {
    const result = await query(
      'DELETE FROM follows WHERE follower_id = $1 AND athlete_id = $2',
      [followerId, athleteId]
    );

    if (result.rowCount === 0) {
      throw new AppError('Not following this athlete', 400);
    }

    logger.info(`User ${followerId} unfollowed athlete ${athleteId}`);
  }

  async getFollowers(athleteId: string, page: number = 1, limit: number = 20): Promise<any> {
    const offset = (page - 1) * limit;

    const result = await query(
      `SELECT u.id, u.username, u.display_name, u.avatar_url, f.created_at as followed_at
       FROM follows f
       JOIN users u ON f.follower_id = u.id
       WHERE f.athlete_id = $1 AND u.is_banned = FALSE
       ORDER BY f.created_at DESC
       LIMIT $2 OFFSET $3`,
      [athleteId, limit, offset]
    );

    const countResult = await query(
      'SELECT COUNT(*) FROM follows WHERE athlete_id = $1',
      [athleteId]
    );

    const total = parseInt(countResult.rows[0].count);

    return {
      followers: result.rows,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  async getFollowing(userId: string, page: number = 1, limit: number = 20): Promise<any> {
    const offset = (page - 1) * limit;

    const result = await query(
      `SELECT u.id, u.username, u.display_name, u.avatar_url,
              ap.is_live, f.created_at as followed_at
       FROM follows f
       JOIN users u ON f.athlete_id = u.id
       JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE f.follower_id = $1 AND u.is_banned = FALSE
       ORDER BY ap.is_live DESC, f.created_at DESC
       LIMIT $2 OFFSET $3`,
      [userId, limit, offset]
    );

    const countResult = await query(
      'SELECT COUNT(*) FROM follows WHERE follower_id = $1',
      [userId]
    );

    const total = parseInt(countResult.rows[0].count);

    return {
      following: result.rows,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  async isFollowing(followerId: string, athleteId: string): Promise<boolean> {
    const result = await query(
      'SELECT 1 FROM follows WHERE follower_id = $1 AND athlete_id = $2',
      [followerId, athleteId]
    );

    return result.rows.length > 0;
  }
}

export const userService = new UserService();

File: backend-go/internal/service/user_service.go
package service

import (
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/phitonias/streak/internal/database"
	"github.com/phitonias/streak/internal/models"
	"gorm.io/gorm"
)

type UserService struct{}

func NewUserService() *UserService {
	return &UserService{}
}

type UpdateProfileInput struct {
	DisplayName     *string
	Bio             *string
	AvatarURL       *string
	BannerURL       *string
	SportCategories []string
}

func (s *UserService) GetUserByID(userID uuid.UUID) (interface{}, error) {
	var user models.User
	result := database.DB.Preload("AthleteProfile").
		Where("id = ? AND is_banned = false", userID).
		First(&user)

	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, result.Error
	}

	return sanitizeUser(&user), nil
}

func (s *UserService) GetUserByUsername(username string) (interface{}, error) {
	var user models.User
	result := database.DB.Preload("AthleteProfile").
		Where("username = ? AND is_banned = false", username).
		First(&user)

	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, result.Error
	}

	return sanitizeUser(&user), nil
}

func (s *UserService) UpdateProfile(userID uuid.UUID, input UpdateProfileInput) (interface{}, error) {
	var user models.User
	if err := database.DB.Where("id = ?", userID).First(&user).Error; err != nil {
		return nil, errors.New("user not found")
	}

	// Update user fields
	updates := make(map[string]interface{})
	if input.DisplayName != nil {
		updates["display_name"] = *input.DisplayName
	}
	if input.Bio != nil {
		updates["bio"] = *input.Bio
	}
	if input.AvatarURL != nil {
		updates["avatar_url"] = *input.AvatarURL
	}
	if input.BannerURL != nil {
		updates["banner_url"] = *input.BannerURL
	}

	if len(updates) > 0 {
		if err := database.DB.Model(&user).Updates(updates).Error; err != nil {
			return nil, err
		}
	}

	// Update athlete profile if sport categories provided
	if len(input.SportCategories) > 0 && user.Role == models.RoleAthlete {
		if err := database.DB.Model(&models.AthleteProfile{}).
			Where("user_id = ?", userID).
			Update("sport_categories", input.SportCategories).Error; err != nil {
			return nil, err
		}
	}

	return s.GetUserByID(userID)
}

func (s *UserService) FollowUser(followerID, athleteID uuid.UUID) error {
	// Check if athlete exists and is an athlete
	var athlete models.User
	if err := database.DB.Where("id = ?", athleteID).First(&athlete).Error; err != nil {
		return errors.New("athlete not found")
	}

	if athlete.Role != models.RoleAthlete {
		return errors.New("can only follow athletes")
	}

	if followerID == athleteID {
		return errors.New("cannot follow yourself")
	}

	// Check if already following
	var existingFollow models.Follow
	result := database.DB.Where("follower_id = ? AND athlete_id = ?", followerID, athleteID).First(&existingFollow)
	if result.Error == nil {
		return errors.New("already following this athlete")
	}

	// Create follow
	follow := models.Follow{
		FollowerID: followerID,
		AthleteID:  athleteID,
	}

	if err := database.DB.Create(&follow).Error; err != nil {
		return err
	}

	// Increment follower count (trigger handles this in DB, but we can also do it here)
	database.DB.Model(&models.AthleteProfile{}).
		Where("user_id = ?", athleteID).
		Update("follower_count", gorm.Expr("follower_count + 1"))

	return nil
}

func (s *UserService) UnfollowUser(followerID, athleteID uuid.UUID) error {
	result := database.DB.Where("follower_id = ? AND athlete_id = ?", followerID, athleteID).
		Delete(&models.Follow{})

	if result.Error != nil {
		return result.Error
	}

	if result.RowsAffected == 0 {
		return errors.New("not following this athlete")
	}

	// Decrement follower count
	database.DB.Model(&models.AthleteProfile{}).
		Where("user_id = ?", athleteID).
		Update("follower_count", gorm.Expr("follower_count - 1"))

	return nil
}

func (s *UserService) GetFollowers(athleteID uuid.UUID, page, limit int) ([]interface{}, *Pagination, error) {
	offset := (page - 1) * limit

	var follows []models.Follow
	if err := database.DB.Preload("Follower").
		Where("athlete_id = ?", athleteID).
		Order("created_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&follows).Error; err != nil {
		return nil, nil, err
	}

	// Get total count
	var total int64
	database.DB.Model(&models.Follow{}).Where("athlete_id = ?", athleteID).Count(&total)

	// Format response
	followers := make([]interface{}, len(follows))
	for i, follow := range follows {
		followers[i] = map[string]interface{}{
			"id":           follow.Follower.ID,
			"username":     follow.Follower.Username,
			"display_name": follow.Follower.DisplayName,
			"avatar_url":   follow.Follower.AvatarURL,
			"followed_at":  follow.CreatedAt,
		}
	}

	pagination := &Pagination{
		Page:       page,
		Limit:      limit,
		Total:      int(total),
		TotalPages: (int(total) + limit - 1) / limit,
	}

	return followers, pagination, nil
}

func (s *UserService) GetFollowing(followerID uuid.UUID, page, limit int) ([]interface{}, *Pagination, error) {
	offset := (page - 1) * limit

	var follows []models.Follow
	if err := database.DB.Preload("Athlete").Preload("Athlete.AthleteProfile").
		Where("follower_id = ?", followerID).
		Order("created_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&follows).Error; err != nil {
		return nil, nil, err
	}

	// Get total count
	var total int64
	database.DB.Model(&models.Follow{}).Where("follower_id = ?", followerID).Count(&total)

	// Format response
	following := make([]interface{}, len(follows))
	for i, follow := range follows {
		isLive := false
		if follow.Athlete.AthleteProfile != nil {
			isLive = follow.Athlete.AthleteProfile.IsLive
		}

		following[i] = map[string]interface{}{
			"id":           follow.Athlete.ID,
			"username":     follow.Athlete.Username,
			"display_name": follow.Athlete.DisplayName,
			"avatar_url":   follow.Athlete.AvatarURL,
			"is_live":      isLive,
			"followed_at":  follow.CreatedAt,
		}
	}

	pagination := &Pagination{
		Page:       page,
		Limit:      limit,
		Total:      int(total),
		TotalPages: (int(total) + limit - 1) / limit,
	}

	return following, pagination, nil
}

func (s *UserService) IsFollowing(followerID, athleteID uuid.UUID) (bool, error) {
	var count int64
	err := database.DB.Model(&models.Follow{}).
		Where("follower_id = ? AND athlete_id = ?", followerID, athleteID).
		Count(&count).Error

	return count > 0, err
}

type Pagination struct {
	Page       int `json:"page"`
	Limit      int `json:"limit"`
	Total      int `json:"total"`
	TotalPages int `json:"totalPages"`
}
